### ISSUE TITLE
Auto-gen anchor doesn't work

### ISSUE THREAD

**Issue by lucasjinreal** [ultralytics/yolov5#117]
Hi, I found that new auto gened anchor actually can not work:

```
Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
   107/299     7.18G   0.04342   0.05105   0.01396    0.1084       655       800: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 809/809 [09:35<00:00,  1.41it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████| 69/69 [00:16<00:00,  4.22it/s]
                 all         548    3.88e+04    0.000376     0.00103    2.35e-06    4.84e-07

```

it's not work

**Comment by glenn-jocher** [comment]
@jinfagang thanks for the feedback. Is there any reason you can think of that it might not have worked correctly? Did the autogenerated anchors appear to correctly fit your labels?

**Comment by glenn-jocher** [comment]
@jinfagang also can you produce your results.png and upload it here?
`from utils.utils import *; plot_results()`

**Comment by glenn-jocher** [comment]
I ran a quick test on a custom dataset myself, with and without autoanchor, and didn't see any problems. In this example the dataset already had good BPR, so I forced autoanchor to run in the second example below.

```
Analyzing anchors... Best Possible Recall (BPR) = 0.9985. Attempting to generate improved anchors, please wait...
WARNING: Extremely small objects found. 360 of 123671 labels are < 4 pixels in width or height.
Running kmeans for 9 anchors on 123611 points...
thr=0.25: 0.9966 best possible recall, 7.46 anchors past thr
n=9, img_size=320, metric_all=0.447/0.770-mean/best, past_thr=0.503-mean: 15,17,  33,23,  20,37,  25,60,  37,44,  58,34,  42,76,  94,61,  54,117
Evolving anchors with Genetic Algorithm: fitness = 0.7832: 100% 1000/1000 [00:51<00:00, 19.39it/s]
thr=0.25: 0.9982 best possible recall, 7.83 anchors past thr
n=9, img_size=320, metric_all=0.472/0.783-mean/best, past_thr=0.514-mean: 13,14,  26,18,  16,30,  27,33,  23,53,  46,28,  37,47,  36,81,  65,57
Original anchors better than new anchors. Proceeding with original anchors.

Image sizes 320 train, 320 test
Using 2 dataloader workers
Starting training for 10 epochs...
```

![results](https://user-images.githubusercontent.com/26833433/84969864-6441b200-b0ce-11ea-9c51-a0bb45c50ecb.png)

EDIT: ah, I see that in my example above the original anchors were used. Ok I'll run a third experiment forcing it to use the new anchors.

**Comment by glenn-jocher** [comment]
Updated results forcing use of autogen anchors. Yes, you are right, there is definitely a problem in the implementation! Let me look through this.

![results (1)](https://user-images.githubusercontent.com/26833433/84970610-e7173c80-b0cf-11ea-8c8e-96eb6ed2f86d.png)

**Comment by lucasjinreal** [comment]
@glenn-jocher  I think the bias happens at test does not use same anchor as training does. I restart traning by pass the auto gened anchor and it seems works now.

**Comment by lucasjinreal** [comment]
I copied the generated anchor and training from scratch again seems ok now. Will test it further to see if there any unnormal.

**Comment by lucasjinreal** [comment]
Why autoanchor_yes and forced_authoanchor doesnt same here? @glenn-jocher

**Comment by glenn-jocher** [comment]
yes_autoanchor(green) runs autoanchor, but the new anchors are not better than the originals, so it uses the original. blue I forced it to use the autoanchor results. I think I know what the problem is. The anchors are used in two places, one set for inference, and a second set for loss computation. The code only updated one set. 

So I think the anchor sizes are correct, the problem is that the loss function is using a different set of anchors. I'm testing a fix now.

**Comment by lucasjinreal** [comment]
@glenn-jocher Same guess. it seems manually replace generated anchor works.

**Comment by glenn-jocher** [comment]
Ok, I've tested a fix, confirmed it works, and pushed a commit with it just now. The updated results are in red. This should fix the bug, but as to whether the results will be better or not than the original anchors is a separate topic, depends really on the dataset. I imagine it will help datasets with low BPRs most. If your BPR is already near 0.99 then the impact may be minimal.

![results (2)](https://user-images.githubusercontent.com/26833433/84972545-eda7b300-b0d3-11ea-88f4-e5028e474692.png)

**Comment by glenn-jocher** [comment]
Fix is in 9fdb0fbacf65ad6d8778da592ef9b87038a3f4c7

**Comment by lucasjinreal** [comment]
@glenn-jocher Red might need more iterations to fit. thanks for your updates, close this issue then.

**Comment by glenn-jocher** [comment]
@jinfagang I followed up on my quick test with a longer training to 100 epochs for both on my custom dataset. The performance between default anchors and autoanchors was close, but the default anchors actually performed better in this case.

I think custom data training may be more complicated than simply fitting the best anchors as I originally though, because the 3 output layers are most sensetive to objects within the size of their receptive fields, which are 8, 16 and 32 pixels for the 3 output layers.

In my example dataset, most objects are very small, about 10 pixels. This means all 9 of the new anchors will be fit to small objects, even the 3 anchors on the 'large' object output layer with stride 32, which may have a very hard time locating very small objects properly.

I think this may require a bit more work to get the best results, perhaps by constraining the anchors to not shrink below their stride size, which is 8, 16 and 32 pixels for the small, medium and large anchors. I'm not sure what the best solution is, I'll have to think about it some more.

![results (3)](https://user-images.githubusercontent.com/26833433/84980896-1fc31000-b0e8-11ea-9b06-15da2613f20e.png)

**Comment by lucasjinreal** [comment]
@glenn-jocher Yes, my dataset is also many tiny obejcts, I found the generated anchors narrow down the sizes then default, this might causes all 3 layer predicted boxes located in same level anchor in some edge cases, this is yolo series algorithms hard to handle. But in my cases, the auto gened anchor actually make the training more faster since the anchor more efficient to guide whole model training.

**Comment by glenn-jocher** [comment]
Removing TODO tag, as AutoAnchor basic functionality is working now, though further refinements will hopefully be applied in the coming months in order to avoid assigning large anchors to small output layers and vice versa.
